pinc/ProofProject.inc

```diff
@@ -0,0 +1,362 @@
+<?php
+declare(strict_types=1);
+
+trait ProofPage
+{
+    private function render_page_data(): array
+    {
+        $round_info_array = [];
+        foreach ($this->round->other_rounds_with_visible_usernames as $other_round_id) {
+            $other_round = get_Round_for_round_id($other_round_id);
+            $username = $this->get_username_for_round($other_round);
+            $round_info_array[] = [$other_round->id, $username];
+        }
+        $page_option = ($this->page_state == $this->round->page_out_state) ? "previous" : "current";
+        return
+            $this->get_page_text_array($page_option) +
+            $this->get_page_state_array() +
+            [
+                "pageNum" => pathinfo($this->page_name, PATHINFO_FILENAME),
+                "roundInfoArray" => $round_info_array,
+                "pagename" => $this->page_name,
+                "image_url" => "{$this->project->url}/{$this->page_name}",
+            ];
+    }
+
+    private function get_username_for_round(object $round): string
+    {
+        $res = DPDatabase::query("
+            SELECT {$round->user_column_name}
+            FROM {$this->projectid}
+            WHERE image = '{$this->page_name}'
+        ");
+        $row = mysqli_fetch_assoc($res);
+        return $row[$round->user_column_name];
+    }
+
+    private function get_page_text_array(string $page_option): array
+    {
+        if ($page_option == "current") {
+            $desired_column_name = $this->round->text_column_name;
+        } else {
+            $desired_column_name = $this->round->prevtext_column_name;
+        }
+
+        $sql = "
+            SELECT $desired_column_name
+            FROM $this->projectid
+            WHERE image='$this->page_name'
+        ";
+        $res = DPDatabase::query($sql);
+        $text = mysqli_fetch_row($res)[0];
+
+        return ["text" => $text];
+    }
+
+    private function get_page_state_array()
+    {
+        return [
+            "pagestate" => $this->page_state,
+            "saved" => ($this->page_state == $this->round->page_temp_state),
+        ];
+    }
+}
+
+class ProofProject
+{
+    use ProofPage;
+
+    public function __construct(object $project, array $query_params = [])
+    {
+        global $pguser, $PROJECT_STATES_IN_ORDER;
+
+        $expected_state = $query_params["projstate"] ?? null;
+        if (null === $expected_state) {
+            throw new ProjectNoStateGivenException("No expected state found in request.");
+        }
+        if (!in_array($expected_state, $PROJECT_STATES_IN_ORDER)) {
+            throw new ProjectInvalidStateException(sprintf("%s is not a valid project state", $expected_state));
+        }
+        if ($expected_state != $project->state) {
+            $err = sprintf(
+                _('Warning: Project "%1$s" is not in state "%2$s"; it is now in state "%3$s".'),
```

Comment:
It may not be consistent but the plan was to localize messages that the user would see in normal use, but not messages that would only appear if there was something wrong with the client application.